activesupport/lib/active_support/core_ext/object/deep_dup.rb

```diff
@@ -43,7 +43,7 @@ class Hash
   def deep_dup
     hash = dup
     each_pair do |key, value|
-      if key.frozen? && ::String === key
+      if ::Symbol === key || (key.frozen? && ::String === key)
```

Comment:
This makes sense to me. It is probably worth running your benchmark with strings as well, just to make sure we haven't made the performance worse there while making it better for symbols (we deep_dup the default Active Record attributes when initializing each record, and those use string keys, so we want to keep that snappy).

You can measure both iterations per second and memory with:

```rb
require 'benchmark/ips'
require 'benchmark/memory'

Benchmark.ips do |bm|
  bm.report("deep_dup") { data.deep_dup }
  bm.report("deep_dup_all") { data.deep_dup_all }
  bm.compare!
end

Benchmark.memory do |bm|
  bm.report("deep_dup") { data.deep_dup }
  bm.report("deep_dup_all") { data.deep_dup_all }
  bm.compare!
end
```

Reply:
neat benchmark extensions!  ips is confirmed to be way up and memory usage also improves dramatically for this PR as is
```
# IPS
Comparison:
        deep_dup_all:       56.8 i/s
             deep_dup:       36.1 i/s - 1.57x  (± 0.00) slower

# Memory
Calculating -------------------------------------
            deep_dup     2.800M memsize (     0.000  retained)
                        17.577k objects (     0.000  retained)
                        50.000  strings (     0.000  retained)
        deep_dup_all     1.621M memsize (     0.000  retained)
                        17.577k objects (     0.000  retained)
                        50.000  strings (     0.000  retained)
Comparison:
        deep_dup_all:    1620640 allocated
             deep_dup:    2800288 allocated - 1.73x more
```

string keys wind up ~8% slower since they come second in the if statement
```
Comparison:
             deep_dup:       47.7 i/s
        deep_dup_all:       44.4 i/s - 1.07x  (± 0.00) slower
```

if we reorder the if statement
```ruby
if (::String === key && key.frozen?) || ::Symbol === key
```
then string performance is the same and symbols perf improves ~30% plus the memory win
```
Comparison:
        deep_dup_all:       51.8 i/s
            deep_dup:       36.1 i/s - 1.44x  (± 0.00) slower
Memory Comparison:
        deep_dup_all:    1620640 allocated
         deep_dup:    2800288 allocated - 1.73x more
```

which do we want to favor slightly more: symbol or string hash keys?

Reply:
actually, the `key.frozen?` check is redundant, since ruby automatically freezes all string keys...so if we remove that check then performance is the same for strings and 35% better for symbols.  that seems like the win, right?

Reply:
> actually, the key.frozen? check is redundant, since ruby automatically freezes all string keys

This is only true for Ruby 2.7+ and Rails 6.1 is supports Ruby 2.6 so this isn't redundant and needs to be added back.

Reply:
@eileencodes gotcha...reverted